File: crates/tmark-ir/src/walk.rs
//! Pre-order traversal of a document.
//!
//! Design: `design/03-ir.md` §Walking and mapping: "matching on the enum is
//! the visitor". Only `walk` is provided; `map` waits for its first user.

use crate::node::{Block, Document, Inline, Row};
use crate::span::{NodeId, Span};

/// A borrowed node handed to a walker.
#[derive(Copy, Clone, Debug)]
pub enum NodeRef<'a> {
    Block(&'a Block),
    Inline(&'a Inline),
}

impl NodeRef<'_> {
    pub fn id(self) -> NodeId {
        match self {
            NodeRef::Block(b) => b.meta().id,
            NodeRef::Inline(i) => i.meta().id,
        }
    }

    pub fn span(self) -> Span {
        match self {
            NodeRef::Block(b) => b.meta().span,
            NodeRef::Inline(i) => i.meta().span,
        }
    }
}

/// Visits every block and inline of the document in pre-order: the body,
/// then the footnote definitions. Table cells, captions, admonition titles,
/// image alt texts and index paths are descended into.
pub fn walk(doc: &Document, f: &mut impl FnMut(NodeRef)) {
    walk_blocks(&doc.blocks, f);
    for footnote in &doc.footnotes {
        walk_blocks(&footnote.content, f);
    }
}

/// Plain text of inlines: the concatenated `Str`, code and math text with
/// breaks as spaces and every other node dropped. Used for table column
/// names, index paths, glossary keys, keystrokes, and outline labels.
pub fn plain_text(inlines: &[Inline]) -> String {
    let mut out = String::new();
    for inline in inlines {
        match inline {
            Inline::Str(s) => out.push_str(&s.text),
            Inline::Space(_) | Inline::SoftBreak(_) | Inline::LineBreak(_) => out.push(' '),
            Inline::Code(c) => out.push_str(&c.text),
            Inline::Math(m) => out.push_str(&m.text),
            Inline::Emph(n) => out.push_str(&plain_text(&n.content)),
            Inline::Strong(n) => out.push_str(&plain_text(&n.content)),
            Inline::Strikeout(n) => out.push_str(&plain_text(&n.content)),
            Inline::Underline(n) => out.push_str(&plain_text(&n.content)),
            Inline::Highlight(n) => out.push_str(&plain_text(&n.content)),
            Inline::Subscript(n) => out.push_str(&plain_text(&n.content)),
            Inline::Superscript(n) => out.push_str(&plain_text(&n.content)),
            Inline::SmallCaps(n) => out.push_str(&plain_text(&n.content)),
            Inline::Quoted(n) => out.push_str(&plain_text(&n.content)),
            Inline::Link(n) => out.push_str(&plain_text(&n.content)),
            Inline::Span(n) => out.push_str(&plain_text(&n.content)),
            Inline::Abbr(a) => out.push_str(&a.text),
            _ => {}
        }
    }
    out
}

pub fn walk_blocks(blocks: &[Block], f: &mut impl FnMut(NodeRef)) {
    for block in blocks {
        walk_block(block, f);
    }
}

pub fn walk_inlines(inlines: &[Inline], f: &mut impl FnMut(NodeRef)) {
    for inline in inlines {
        walk_inline(inline, f);
    }
}

fn walk_block(block: &Block, f: &mut impl FnMut(NodeRef)) {
    f(NodeRef::Block(block));
    match block {
        Block::Para(n) => {
            if let Some(lead) = &n.lead {
                walk_inlines(lead, f);
            }
            walk_inlines(&n.content, f);
        }
        Block::Plain(n) => walk_inlines(&n.content, f),
        Block::Header(n) => walk_inlines(&n.content, f),
        Block::BlockQuote(n) => walk_blocks(&n.content, f),
        Block::BulletList(n) => {
            for item in &n.items {
                walk_blocks(&item.content, f);
            }
        }
        Block::OrderedList(n) => {
            for item in &n.items {
                walk_blocks(&item.content, f);
            }
        }
        Block::DefinitionList(n) => {
            for (term, definitions) in &n.items {
                walk_inlines(term, f);
                for definition in definitions {
                    walk_blocks(definition, f);
                }
            }
        }
        Block::Table(n) => {
            for row in n.model.rows.iter().chain(&n.model.footer) {
                if let Row::Data(row) = row {
                    for cell in &row.cells {
                        walk_inlines(&cell.content, f);
                    }
                }
            }
        }
        Block::Caption(n) => walk_inlines(&n.content, f),
        Block::Figure(n) => walk_blocks(&n.content, f),
        Block::Admonition(n) => {
            if let Some(title) = &n.title {
                walk_inlines(title, f);
            }
            walk_blocks(&n.content, f);
        }
        Block::Div(n) => walk_blocks(&n.content, f),
        Block::CodeBlock(_)
        | Block::HorizontalRule(_)
        | Block::TableConfig(_)
        | Block::MathBlock(_)
        | Block::RawBlock(_)
        | Block::Include(_)
        | Block::Comment(_) => {}
    }
}

fn walk_inline(inline: &Inline, f: &mut impl FnMut(NodeRef)) {
    f(NodeRef::Inline(inline));
    match inline {
        Inline::Emph(n) => walk_inlines(&n.content, f),
        Inline::Strong(n) => walk_inlines(&n.content, f),
        Inline::Strikeout(n) => walk_inlines(&n.content, f),
        Inline::Underline(n) => walk_inlines(&n.content, f),
        Inline::Highlight(n) => walk_inlines(&n.content, f),
        Inline::Subscript(n) => walk_inlines(&n.content, f),
        Inline::Superscript(n) => walk_inlines(&n.content, f),
        Inline::SmallCaps(n) => walk_inlines(&n.content, f),
        Inline::Quoted(n) => walk_inlines(&n.content, f),
        Inline::Link(n) => walk_inlines(&n.content, f),
        Inline::Note(n) => walk_blocks(&n.content, f),
        Inline::Image(n) => walk_inlines(&n.alt, f),
        Inline::IndexEntry(n) => {
            for level in &n.path {
                walk_inlines(level, f);
            }
        }
        Inline::Aside(n) => walk_blocks(&n.content, f),
        Inline::Span(n) => walk_inlines(&n.content, f),
        Inline::Str(_)
        | Inline::Space(_)
        | Inline::SoftBreak(_)
        | Inline::LineBreak(_)
        | Inline::Code(_)
        | Inline::Math(_)
        | Inline::Ref(_)
        | Inline::CounterItem(_)
        | Inline::Keystroke(_)
        | Inline::Var(_)
        | Inline::Abbr(_)
        | Inline::Comment(_)
        | Inline::RawInline(_) => {}
    }
}

File: crates/tmark-py/python/tmark/ir/walk.py
"""Traversal utilities over the generated IR (:mod:`tmark.ir.model`).

Same entry points as :mod:`tmark.ir.visitor` has for the legacy tree, over
the tmark-shaped models:

* :func:`walk` — every node in pre-order, in tmark's order (``walk.rs``): a
  block before its inlines, ``Para.lead`` before ``content``, an admonition
  title before its body, table rows then footer rows, the document body
  then the footnote definitions.
* :func:`children`, :func:`iter_child_fields` — direct children, found
  structurally: any dataclass field holding a :class:`~tmark.ir.model.Node`,
  a tuple of them, or a :class:`~tmark.ir.model.Record` that holds them
  (list items, table cells, footnotes).
* :func:`map_tree` — rebuild bottom-up; ``fn`` sees nodes only, records and
  tuples are rebuilt around them and unchanged subtrees are reused.
* :func:`map_inlines` — rebuild the inline lists: ``fn`` may return one
  inline or a tuple of them (spliced into the parent's list), and ``skip``
  fences off subtrees (a code span, an emoji span) from the rewrite.
* :class:`NodeVisitor` — ``visit_<ClassName>`` double dispatch along the MRO.
* :func:`plain_text` — tmark's ``plain_text``: the text of ``Str``, ``Code``,
  ``Math`` and ``Abbr``, breaks as spaces, formatting containers and links
  descended into, everything else dropped.
"""

from __future__ import annotations

from collections.abc import Callable, Iterable, Iterator
from dataclasses import fields, replace
from typing import Any, TypeVar

from tmark.ir import model
from tmark.ir.model import Inline, Node, Record


__all__ = [
    "NodeVisitor",
    "children",
    "iter_child_fields",
    "map_inlines",
    "map_tree",
    "plain_text",
    "walk",
]

T = TypeVar("T", bound=Node | Record)


def _holds_nodes(value: Any) -> bool:
    """True if ``value`` is a node or contains one (tuples and records descended)."""
    if isinstance(value, Node):
        return True
    if isinstance(value, tuple):
        return any(_holds_nodes(item) for item in value)
    if isinstance(value, Record):
        return any(_holds_nodes(getattr(value, f.name)) for f in fields(value))
    return False


#: Where a type's fields are not visited in declaration order. The dataclasses
#: are generated from the JSON schema, whose properties schemars sorts
#: alphabetically, so the order of ``TableModel`` says nothing; ``tmark_ir::walk``
#: visits a table's column titles (the header row) before its body rows.
_CHILD_ORDER: dict[str, tuple[str, ...]] = {
    "TableModel": ("columns", "rows", "footer"),
    "Para": ("lead", "content"),
}


def _ordered_fields(node: Node | Record) -> tuple[str, ...]:
    names = tuple(f.name for f in fields(node))
    order = _CHILD_ORDER.get(type(node).__name__)
    if order is None:
        return names
    return order + tuple(name for name in names if name not in order)


def iter_child_fields(node: Node | Record) -> Iterator[tuple[str, Any]]:
    """Yield ``(field_name, value)`` for each field holding child node(s).

    Scalar fields, enums, spans and records without nodes (``Attrs``, the
    front matter, ``TableSettings``) are skipped. The order is tmark's
    pre-order (see :data:`_CHILD_ORDER`), not the declaration order.
    """
    for name in _ordered_fields(node):
        value = getattr(node, name)
        if _holds_nodes(value):
            yield name, value


def _iter_nodes(value: Any) -> Iterator[Node]:
    """The nodes directly inside ``value``, through tuples and records."""
    if isinstance(value, Node):
        yield value
    elif isinstance(value, tuple):
        for item in value:
            yield from _iter_nodes(item)
    elif isinstance(value, Record):
        for name in _ordered_fields(value):
            yield from _iter_nodes(getattr(value, name))


def children(node: Node | Record) -> tuple[Node, ...]:
    """The direct child nodes of ``node``, in tmark's pre-order."""
    result: list[Node] = []
    for name in _ordered_fields(node):
        result.extend(_iter_nodes(getattr(node, name)))
    return tuple(result)


def walk(root: Node | Record) -> Iterator[Node]:
    """Yield ``root`` (when it is a node) then every descendant node, pre-order.

    A :class:`~tmark.ir.model.Document` is not a node: ``walk(doc)`` yields
    its blocks and their descendants, then the footnote contents, as tmark's
    ``walk`` does.
    """
    if isinstance(root, Node):
        yield root
    for child in children(root):
        yield from walk(child)


def _map_value(value: Any, fn: Callable[[Node], Node]) -> Any:
    if isinstance(value, Node):
        return _map_node(value, fn)
    if isinstance(value, tuple):
        items = tuple(_map_value(item, fn) for item in value)
        return value if all(a is b for a, b in zip(items, value, strict=True)) else items
    if isinstance(value, Record):
        return _map_record(value, fn)
    return value


def _map_fields(obj: T, fn: Callable[[Node], Node]) -> T:
    changes: dict[str, Any] = {}
    for name, value in iter_child_fields(obj):
        mapped = _map_value(value, fn)
        if mapped is not value:
            changes[name] = mapped
    return replace(obj, **changes) if changes else obj


def _map_record(record: T, fn: Callable[[Node], Node]) -> T:
    return _map_fields(record, fn)


def _map_node(node: Node, fn: Callable[[Node], Node]) -> Node:
    return fn(_map_fields(node, fn))


def map_tree(root: T, fn: Callable[[Node], Node]) -> T:
    """Return a new tree with ``fn`` applied to every node, bottom-up.

    Children are transformed before their parent, so ``fn`` sees already
    mapped descendants. Nothing is mutated; a subtree ``fn`` returns
    unchanged is reused. ``id`` and ``span`` travel with ``replace``.
    """
    return _map_value(root, fn)


InlineMap = Callable[[Inline], "Inline | tuple[Inline, ...]"]


def _map_inline_value(value: Any, fn: InlineMap, skip: Callable[[Node], bool]) -> Any:
    if isinstance(value, Node):
        if skip(value):
            return value
        rebuilt = _map_inline_fields(value, fn, skip)
        return fn(rebuilt) if isinstance(rebuilt, Inline) else rebuilt
    if isinstance(value, tuple):
        items: list[Any] = []
        changed = False
        for item in value:
            mapped = _map_inline_value(item, fn, skip)
            if isinstance(item, Inline) and isinstance(mapped, tuple):
                items.extend(mapped)
                changed = True
                continue
            if mapped is not item:
                changed = True
            items.append(mapped)
        return tuple(items) if changed else value
    if isinstance(value, Record):
        return _map_inline_fields(value, fn, skip)
    return value


def _map_inline_fields(obj: T, fn: InlineMap, skip: Callable[[Node], bool]) -> T:
    changes: dict[str, Any] = {}
    for name, value in iter_child_fields(obj):
        mapped = _map_inline_value(value, fn, skip)
        if mapped is not value:
            changes[name] = mapped
    return replace(obj, **changes) if changes else obj


def _never(node: Node) -> bool:
    del node
    return False


def map_inlines(
    root: T,
    fn: InlineMap,
    *,
    skip: Callable[[Node], bool] = _never,
) -> T:
    """Return a new tree with ``fn`` applied to every inline, bottom-up.

    Like :func:`map_tree` but for the inline lists: ``fn`` sees every
    :class:`~tmark.ir.model.Inline` (children first) and returns either a
    replacement inline or a tuple of inlines, which is spliced in place of the
    node in its parent's list (an empty tuple drops it). Blocks and records
    are rebuilt around the changes; ``skip(node)`` fences off a subtree — a
    node it accepts is kept as is, its descendants unseen. Unchanged subtrees
    are reused, ``id`` and ``span`` travel with ``replace``.
    """
    return _map_inline_value(root, fn, skip)


class NodeVisitor:
    """Type-dispatching visitor over the IR.

    Define ``visit_<ClassName>(self, node)`` for the classes you handle;
    :meth:`visit` resolves the method along the node's MRO, so ``visit_Block``
    or ``visit_Inline`` catches a family. Anything unmatched reaches
    :meth:`generic_visit`, which visits the children and returns ``None``.
    """

    def visit(self, node: Node | Record) -> Any:
        """Dispatch to the most specific ``visit_<ClassName>`` for ``node``."""
        for klass in type(node).__mro__:
            method = getattr(self, f"visit_{klass.__name__}", None)
            if method is not None:
                return method(node)
        return self.generic_visit(node)

    def generic_visit(self, node: Node | Record) -> Any:
        """Default: visit each child. Override to customise the fallback."""
        for child in children(node):
            self.visit(child)
        return None


_TEXT = (model.Str, model.Code, model.Math, model.Abbr)
_BREAKS = (model.Space, model.SoftBreak, model.LineBreak)
_THROUGH = (
    model.Emph,
    model.Strong,
    model.Strikeout,
    model.Underline,
    model.Highlight,
    model.Subscript,
    model.Superscript,
    model.SmallCaps,
    model.Quoted,
    model.Link,
    model.SpanNode,
)


def plain_text(inlines: Iterable[Inline]) -> str:
    """The plain text of ``inlines``, as tmark's ``plain_text`` (``walk.rs``).

    ``Str``, ``Code``, ``Math`` and ``Abbr`` contribute their text; ``Space``,
    ``SoftBreak`` and ``LineBreak`` a space; emphasis-like containers, quotes,
    links and spans their content; every other node nothing.
    """
    out: list[str] = []
    for inline in inlines:
        if isinstance(inline, _TEXT):
            out.append(inline.text)
        elif isinstance(inline, _BREAKS):
            out.append(" ")
        elif isinstance(inline, _THROUGH):
            out.append(plain_text(inline.content))
    return "".join(out)
